lib/Frontend/FrontendOptions.cpp

```diff
@@ -211,3 +198,254 @@ bool FrontendOptions::isOutputFileDirectory() const {
          llvm::sys::fs::is_directory(getSingleOutputFilename());
 }
 
+const char *FrontendOptions::computeSuffix() {
+  switch (RequestedAction) {
+    case NoneAction:
+      return nullptr;
+      
+    case Parse:
+    case Typecheck:
+    case DumpParse:
+    case DumpInterfaceHash:
+    case DumpAST:
+    case EmitSyntax:
+    case PrintAST:
+    case DumpScopeMaps:
+    case DumpTypeRefinementContexts:
+      // Textual modes.
+      return nullptr;
+      
+    case EmitPCH:
+      return PCH_EXTENSION;
+      
+    case EmitSILGen:
+    case EmitSIL:
+      return SIL_EXTENSION;
+      
+    case EmitSIBGen:
+    case EmitSIB:
+      return SIB_EXTENSION;
+      
+    case MergeModules:
+    case EmitModuleOnly:
+      return SERIALIZED_MODULE_EXTENSION;
+      
+    case Immediate:
+    case REPL:
+      // These modes have no frontend-generated output.
+      return nullptr;
+      
+    case EmitAssembly:
+      return "s";
+      
+    case EmitIR:
+      return "ll";
+      
+    case EmitBC:
+      return "bc";
+      
+    case EmitObject:
+      return "o";
+      
+    case EmitImportedModules:
+      return "importedmodules";
+  }
+}
+
+bool FrontendOptions::canEmitDependencies() const {
```

Comment:
I feel like these shouldn't be instance methods; they should just take an ActionType and return the right answer. (They're really properties of the action, but C++ won't let us make them methods of an enum.)